### PR TITLE
[FLINK-30052][Connectors/Kafka] Move existing Kafka connector code from Flink repo to dedicated Kafka repo

### PR DESCRIPTION
**Pull request by mas-chen**
# What is the purpose of the change
Move existing Kafka connector code from Flink repo to dedicated Kafka repo for 1.16 release branch.

# Brief change log
* Move docs, datastream connectors, sql connectors, and e2e tests.
* Create new maven projects and resolve dependency converge issues
* Refactor e2e test code depending on `flink-end-to-end-tests-common`

# Verifying this change
It can be verified by checking if "mvn clean package" works successfully.

### CODE REVIEW

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientKafkaITCase.java [hunk 1]
```diff
@@ -82,15 +86,27 @@ private static Configuration getConfiguration() {
         return flinkConfig;
     }
 
-    @Rule
```
Comment (mas-chen) [inline flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientKafkaITCase.java]: Refactored to remove the `flink-end-to-end-tests-common` dependency

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientSchemaRegistryITCase.java [hunk 2]
```diff
@@ -68,7 +68,8 @@ public class SQLClientSchemaRegistryITCase {
     private static final Path sqlAvroJar = ResourceTestUtils.getResource(".*avro.jar");
     private static final Path sqlAvroRegistryJar =
             ResourceTestUtils.getResource(".*avro-confluent.jar");
-    private static final Path sqlToolBoxJar = ResourceTestUtils.getResource(".*SqlToolbox.jar");
+    private static final Path sqlToolBoxJar =
```
Comment (mas-chen) [inline flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientSchemaRegistryITCase.java:71]: Refactored to remove the `flink-end-to-end-tests-common` dependency

File: flink-connector-kafka/pom.xml [hunk 3]
```diff
@@ -223,6 +228,35 @@ under the License.
 			<artifactId>flink-architecture-tests-test</artifactId>
 			<scope>test</scope>
 		</dependency>
+
+		<dependency>
```
Comment (mas-chen) [inline flink-connector-kafka/pom.xml:232]: A few junit4 tests remain

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+
+		<jackson-bom.version>2.13.4.20221013</jackson-bom.version>
+		<junit4.version>4.13.2</junit4.version>
+		<junit5.version>5.8.1</junit5.version>
+		<assertj.version>3.23.1</assertj.version>
+		<archunit.version>0.22.0</archunit.version>
+		<testcontainers.version>1.17.2</testcontainers.version>
+		<mockito.version>3.4.6</mockito.version>
+		<hamcrest.version>1.3</hamcrest.version>
+		<byte-buddy.version>1.12.10</byte-buddy.version>
```
Comment (mas-chen) [inline pom.xml:65]: took these versions from 1.16 release branch

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/main/java/org/apache/flink/tests/util/kafka/LocalStandaloneKafkaResourceFactory.java [hunk 5]
```diff
@@ -1,48 +0,0 @@
...
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import java.nio.file.Path;
-import java.nio.file.Paths;
-import java.util.Optional;
-
-/** A {@link KafkaResourceFactory} for the {@link LocalStandaloneKafkaResourceFactory}. */
-public final class LocalStandaloneKafkaResourceFactory implements KafkaResourceFactory {
```
Comment (mas-chen) [inline flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/main/java/org/apache/flink/tests/util/kafka/LocalStandaloneKafkaResourceFactory.java]: Refactored to remove the `flink-end-to-end-tests-common` dependency. Using the new testcontainer utilities

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/main/java/org/apache/flink/tests/util/kafka/LocalStandaloneKafkaResource.java [hunk 6]
```diff
@@ -1,452 +0,0 @@
...
-import java.util.concurrent.atomic.AtomicReference;
-import java.util.function.Consumer;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
-
-/**
- * {@link KafkaResource} that downloads kafka and sets up a local kafka cluster with the bundled
- * zookeeper.
- */
-public class LocalStandaloneKafkaResource implements KafkaResource {
```
Comment (mas-chen) [inline flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/main/java/org/apache/flink/tests/util/kafka/LocalStandaloneKafkaResource.java]: Refactored to remove the `flink-end-to-end-tests-common` dependency. Using the new testcontainer utilities

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/main/java/org/apache/flink/tests/util/kafka/KafkaResourceFactory.java [hunk 7]
```diff
@@ -1,36 +0,0 @@
...
- * limitations under the License.
- */
-
-package org.apache.flink.tests.util.kafka;
-
-import java.util.Optional;
-
-/** A factory for {@link KafkaResource} implementations. */
-@FunctionalInterface
-public interface KafkaResourceFactory {
```
Comment (mas-chen) [inline flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/main/java/org/apache/flink/tests/util/kafka/KafkaResourceFactory.java]: Refactored to remove the `flink-end-to-end-tests-common` dependency. Using the new testcontainer utilities

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/main/java/org/apache/flink/tests/util/kafka/KafkaResource.java [hunk 8]
```diff
@@ -1,131 +0,0 @@
...
-import org.apache.flink.tests.util.util.FactoryUtils;
-import org.apache.flink.util.ExternalResource;
-
-import java.io.IOException;
-import java.net.InetSocketAddress;
-import java.util.Collection;
-import java.util.List;
-
-/** Generic interface for interacting with Kafka. */
-public interface KafkaResource extends ExternalResource {
```
Comment (mas-chen) [inline flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/main/java/org/apache/flink/tests/util/kafka/KafkaResource.java]: Refactored to remove the `flink-end-to-end-tests-common` dependency. Using the new testcontainer utilities

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -80,7 +80,7 @@ jobs:
         run: wget -q -c ${{ inputs.flink_url }} -O - | tar -xz
 
       - name: Compile and test flink-connector-kafka
-        timeout-minutes: 30
+        timeout-minutes: 45
```
Comment (mas-chen) [inline .github/workflows/ci.yml:83]: e2e tests were timing out

**Comment by MartijnVisser** [comment]
@mas-chen Thanks a lot for the PR! I've approved the CI run, so we can see if successfully builds.

**Comment by mas-chen** [comment]
@MartijnVisser CI is passing, sorry committed something unintentionally

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/pom.xml [hunk 10]
```diff
@@ -18,230 +18,230 @@ specific language governing permissions and limitations
...
+    <parent>
+        <groupId>org.apache.flink</groupId>
+        <artifactId>flink-connector-kafka-e2e-tests</artifactId>
+        <version>3.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
 
-	<artifactId>flink-end-to-end-tests-common-kafka</artifactId>
-	<name>Flink : E2E Tests : Common Kafka</name>
+    <artifactId>flink-end-to-end-tests-common-kafka</artifactId>
```
Comment (zentol) [inline flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/pom.xml:30]: Inconsistent indentation with remaining poms (tabs vs spaces)

File: pom.xml [hunk 11]
```diff
@@ -0,0 +1,506 @@
...
+		<dependency>
+			<groupId>org.apache.logging.log4j</groupId>
+			<artifactId>log4j-core</artifactId>
+			<scope>test</scope>
+		</dependency>
+
+		<dependency>
+			<!-- API bridge between log4j 1 and 2 -->
+			<groupId>org.apache.logging.log4j</groupId>
+			<artifactId>log4j-1.2-api</artifactId>
```
Comment (zentol) [inline pom.xml:154]: Is any dependency from the kafka connector using log4j 1?
if so, which? (and please add a comment for that)
if not, then we can remove this.

File: pom.xml [hunk 12]
```diff
@@ -0,0 +1,506 @@
...
+				<exclusions>
+					<exclusion>
+						<groupId>log4j</groupId>
+						<artifactId>log4j</artifactId>
+					</exclusion>
+					<exclusion>
+						<groupId>org.slf4j</groupId>
+						<artifactId>slf4j-log4j12</artifactId>
+					</exclusion>
+				</exclusions>
```
Comment (zentol) [inline pom.xml:265]: No longer required since 1.16.0

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientKafkaITCase.java [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+import static org.hamcrest.Matchers.arrayContainingInAnyOrder;
+import static org.junit.Assert.assertThat;
+
+/** End-to-end test for the kafka SQL connectors. */
+@RunWith(Parameterized.class)
+@Ignore("FLINK-21796")
```
Comment (zentol) [inline flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientKafkaITCase.java:70]:
```suggestion
```
Let's see if its still unstable.

**Review by zentol — CHANGES_REQUESTED** [review]
Missing production architecture tests as per `Lack of production architecture tests` in https://cwiki.apache.org/confluence/display/FLINK/Externalized+Connector+development.

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/pom.xml [hunk 14]
```diff
@@ -18,230 +18,230 @@ specific language governing permissions and limitations
...
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>8</source>
+                    <target>8</target>
+                </configuration>
+            </plugin>
```
[the remark this answers is not in the document]
Comment (mas-chen) [reply flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/pom.xml]: This must have been a copy paste error

File: flink-sql-connector-kafka/pom.xml [hunk 15]
```diff
@@ -24,20 +24,16 @@ under the License.
...
+		<version>3.0-SNAPSHOT</version>
 	</parent>
 
 	<artifactId>flink-sql-connector-kafka</artifactId>
 	<name>Flink : Connectors : SQL : Kafka</name>
 
 	<packaging>jar</packaging>
 
-	<properties>
-		<japicmp.skip>true</japicmp.skip>
```
[the remark this answers is not in the document]
Comment (mas-chen) [reply flink-sql-connector-kafka/pom.xml]: Reverted

File: pom.xml (hunk 11 above)
Comment (mas-chen) [reply pom.xml]: No dependency uses it, removed

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/pom.xml (hunk 10 above)
Comment (mas-chen) [reply flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/pom.xml:30]: I fixed all other pots to use spaces

**Comment by zentol** [comment]
hmm, squashed the commits and made some docs changes and now the e2e tests are failing :/

**Comment by mas-chen** [comment]
@zentol It's failing after I un-ignored the e2e test. There's an issue related with initializing the Docker image--I'm actively looking at it

File: .github/workflows/ci.yml (hunk 9 above)
Comment (MartijnVisser) [reply .github/workflows/ci.yml:83]: You should rebase, so this file becomes obsolete (we're re-using the same CI setup across the different externalized connectors). Timeouts are there set to 50 mins for this step and 60 mins for the entire workflow.

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientKafkaITCase.java (hunk 1 above)
Comment (mas-chen) [reply flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientKafkaITCase.java]: I actually propose to get rid of this test. Reason being is that SQLClientSchemaRegistryITCase tests SQL Kafka table source and sink already.

This test is just testing E2E the json format -> avro format -> csv format via file sink. However, now that we are using containerized Flink, the file it writes to is now within the container and it is difficult to obtain the file from the taskmanager containers to assert on the file output contents.

We could easily fix this test if we can use LocalStandaloneFlinkResourceFactory (but it is in flink-end-to-end-test-common). 

Plus I had to create an extraneous module to get the jar of a custom sql function (see flink-sql-toolbox module). 

Alternatively, I can totally rewrite and simplify the test. (only testing json format -> avro format and asserting Kafka output contents, but this is what SQLClientSchemaRegistryITCase is essentially doing).

@zentol @MartijnVisser WDYT?

Comment (mas-chen) [reply flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientKafkaITCase.java]: (there was previous feedback to try un-ignoring this test)

File: .github/workflows/ci.yml (hunk 9 above)
Comment (mas-chen) [reply .github/workflows/ci.yml]: Nice, removed!

File: flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientKafkaITCase.java (hunk 1 above)
Comment (mas-chen) [reply flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientKafkaITCase.java]: @MartijnVisser are we just waiting for Chesnay to come back from vacation to continue this conversation? I had wanted to get this connector ready before the 1.16.1 release. This is the only test that is failing in CI

Comment (mas-chen) [reply flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientKafkaITCase.java]: @zentol @MartijnVisser hope you enjoyed your break! Do you have to time to revisit this discussion^?

It's been a while so just reminding--this is a e2e test that was previously @Ignore'ed and you left a comment to try and re-enable it since the commented JIRA ticket that was blocking this test was fixed. It's possible to support this with some improvements of the Flink testcontainer framework (I can go into the details of what is required). But I think the E2E test is redundant as per my earlier comment on Dec 15.

Comment (zentol) [reply flink-connector-kafka-e2e-tests/flink-end-to-end-tests-common-kafka/src/test/java/org/apache/flink/tests/util/kafka/SQLClientKafkaITCase.java]: I'd be fine with dropping this test.

**Comment by mas-chen** [comment]
CI now passes on my fork

File: flink-connector-kafka/pom.xml [hunk 16]
```diff
@@ -18,259 +18,292 @@ specific language governing permissions and limitations
...
-
-	<modelVersion>4.0.0</modelVersion>
-
-	<parent>
-		<artifactId>flink-connectors</artifactId>
-		<groupId>org.apache.flink</groupId>
-		<version>1.16-SNAPSHOT</version>
-	</parent>
-
-	<artifactId>flink-connector-kafka</artifactId>
```
Comment (zentol) [inline flink-connector-kafka/pom.xml]: can we make the whitespace changes in a separate commit?

**Review by zentol — APPROVED** [review]
Beyond 1 comment about commit structure I think we can go ahead with this one.

**Comment by mas-chen** [comment]
Thanks for your help @zentol @MartijnVisser!
